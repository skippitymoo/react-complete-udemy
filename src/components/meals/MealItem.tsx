import React, { JSX, useState } from 'react';
import { formatCurrency } from '../../utils/utilities';
import { Input } from '../ui/Input';
import { Button } from '../ui/Button';
import { Meal } from './meal.types';

const DEFAULT_ADD_TO_BASKET_AMOUNT = 1;

export type AddToBasket = {
  /**
   * Receives the number of items added by the user
   */
  onAddToBasket: (numberOfItems: number) => void;
};

const MealItemComponent = ({ children }: React.HTMLAttributes<HTMLDivElement>): JSX.Element => {
  return <div className='meal-item'>{children}</div>;
};

const MealItemDetails = ({ name, description, price }: Omit<Meal, 'id'>): JSX.Element => {
  return (
    <div className='meal-item__details'>
      <h3 className='meal-item__title'>{name}</h3>
      <div className='meal-item__description'>{description}</div>
      <div className='meal-item__price'>{formatCurrency(price)}</div>
    </div>
  );
};

const MealItemAddToBasket = ({ onAddToBasket }: AddToBasket): JSX.Element => {
  const [enteredAmount, setEnteredAmount] = useState(DEFAULT_ADD_TO_BASKET_AMOUNT);

  const submitHandler = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();

    onAddToBasket(enteredAmount);
  };

  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>): undefined => {
    setEnteredAmount(+event.target.value);
  };

  return (
    <form name='Add meal item' className='meal-item__add' onSubmit={submitHandler}>
      <Input label='Amount' value={enteredAmount} type='number' onChange={changeHandler} />
      <Button size='long'>+ Add</Button>
    </form>
  );
};

MealItemComponent.Details = MealItemDetails;
MealItemComponent.AddToBasket = MealItemAddToBasket;

export const MealItem = MealItemComponent;
