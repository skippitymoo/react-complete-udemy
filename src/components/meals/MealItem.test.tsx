import { render } from '@testing-library/react';
import { MealItem } from './MealItem';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);

describe('MealItem', () => {
  it('renders the MealItem component and its children', async () => {
    const { container } = render(
      <MealItem>
        <MealItem.Details title='item title' description='item description' price={22.99} />
      </MealItem>,
    );

    expect(await axe(container)).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });
});
