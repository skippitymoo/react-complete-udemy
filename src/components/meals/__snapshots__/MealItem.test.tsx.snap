// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealItem renders the MealItem component and its children 1`] = `
<div
  class="meal-item"
>
  <div
    class="meal-item__details"
  >
    <h3
      class="meal-item__title"
    >
      item title
    </h3>
    <div
      class="meal-item__description"
    >
      item description
    </div>
    <div
      class="meal-item__price"
    >
      Â£22.99
    </div>
  </div>
  <form
    class="meal-item__add"
    name="Add meal item"
  >
    <div
      class="form__group"
    >
      <label
        class="form__label"
        for=":r0:"
      >
        Amount
      </label>
      <input
        class="form__input"
        id=":r0:"
        type="number"
        value="1"
      />
    </div>
    <button
      class="btn btn--primary btn--long"
    >
      + Add
    </button>
  </form>
</div>
`;
